packages/flutter/lib/src/gestures/mouse_tracking.dart

```diff
@@ -104,60 +129,77 @@ class MouseTracker extends ChangeNotifier {
     _router.removeGlobalRoute(_handleEvent);
   }
 
-  // The pointer router that the mouse tracker listens to for events.
-  final PointerRouter _router;
-
-  /// Used to find annotations at a given logical coordinate.
+  /// Find annotations at a given offset in global logical coordinate space.
+  ///
+  /// [MouseTracker] uses this callback to know which annotations are affected
+  /// by each device.
   final MouseDetectorAnnotationFinder annotationFinder;
 
-  // The collection of annotations that are currently being tracked. They may or
-  // may not be active, depending on the value of _TrackedAnnotation.active.
-  final Map<MouseTrackerAnnotation, _TrackedAnnotation> _trackedAnnotations = <MouseTrackerAnnotation, _TrackedAnnotation>{};
+  // The pointer router that the mouse tracker listens to, and receives new
+  // mouse events from.
+  final PointerRouter _router;
 
-  /// Track an annotation so that if the mouse enters it, we send it events.
-  ///
-  /// This is typically called when the [AnnotatedRegion] containing this
-  /// annotation has been added to the layer tree.
-  void attachAnnotation(MouseTrackerAnnotation annotation) {
-    _trackedAnnotations[annotation] = _TrackedAnnotation(annotation);
-    // Schedule a check so that we test this new annotation to see if any mouse
-    // is currently inside its region. It has to happen after the frame is
-    // complete so that the annotation layer has been added before the check.
-    if (mouseIsConnected) {
-      _scheduleMousePositionCheck();
+  // Tracks the state of connected mouse devices.
+  //
+  // It is a source-of-truth for the list of connected mouse devices.
+  //
+  // If a device disconnects, the device will be moved to
+  // `_disconnectedMouseState`.
+  final Map<int, _MouseState> _mouseStates = <int, _MouseState>{};
+
+  // Tracks the state of the disconnected mouse device before its exit callback
+  // is called.
+  //
+  // A device is moved from `_mouseStates` on its disconnecting event, and
+  // will be permanantly removed after its exit callback.
```

Comment:
I don't understand why you need this field at all. Can you not just mark a mouse as disconnected in its _MouseState and then remove it from _mouseStates after sendMouseNotifications has been called in _removeMouseDevice?

Reply:
It was indeed the initial design, but then I discovered that there can only be one disconnected mouse state at a given time, and even that is temporary that only lasts during 2 lines within `_removeMouseDevice`. Therefore it doesn't really worth assigning another field for all mouse states.
I plan to move it as a parameter of `sendMouseNotifications` after the method becomes private, so that `_disconnectedMouseState` can be removed completely, but for now I think this is the easiest plan.